server/cluster/cluster.go

```diff
@@ -0,0 +1,338 @@
+		}
+		shardTables[shardID] = &ShardTablesWithRole{shardRole: shardRole, tables: tables, version: shardTable.version}
+	}
+
+	return shardTables, nil
+}
+
+func (c *Cluster) DropTable(ctx context.Context, schemaName, tableName string, tableID uint64) error {
+	schema, exists := c.getSchema(schemaName)
+	if exists {
+		return ErrSchemaNotFound.WithCausef("schemaName:%s", schemaName)
+	}
+	if err := c.storage.DeleteTables(ctx, c.clusterID, schema.GetID(), []uint64{tableID}); err != nil {
+		return errors.Wrapf(err, "cluster DropTable, "+
+			"clusterID:%d, schema:%v, tableID:%d", c.clusterID, schema, tableID)
+	}
+	c.Lock()
+	defer c.Unlock()
+	schema.dropTableLocked(tableName)
+	for _, shard := range c.shards {
+		shard.dropTableLocked(tableID)
+	}
+	return nil
+}
+
+func (c *Cluster) CreateSchema(ctx context.Context, schemaName string, schemaID uint32) (*Schema, error) {
+	c.Lock()
+	defer c.Unlock()
+	// check if exists
+	{
+		schema, exists := c.getSchema(schemaName)
+		if exists {
+			return schema, nil
+		}
+	}
+
+	// persist
+	schemaPb := &clusterpb.Schema{Id: schemaID, Name: schemaName, ClusterId: c.clusterID}
+	err := c.storage.CreateSchema(ctx, c.clusterID, schemaPb)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster CreateSchema")
+	}
+
+	// cache
+	schema := c.updateSchemaCacheLocked(schemaPb)
+	return schema, nil
+}
+
+func (c *Cluster) CreateTable(ctx context.Context, schemaName string, shardID uint32,
+	tableName string, tableID uint64,
+) (*Table, error) {
+	c.Lock()
+	defer c.Unlock()
+	// check if exists
+	schema, exists := c.getSchema(schemaName)
+	if !exists {
+		return nil, ErrSchemaNotFound.WithCausef("schemaName", schemaName)
+	}
+
+	// persist
+	tablePb := &clusterpb.Table{Id: tableID, Name: tableName, SchemaId: schema.GetID(), ShardId: shardID}
+	err1 := c.storage.CreateTable(ctx, c.clusterID, schema.GetID(), tablePb)
+	if err1 != nil {
+		return nil, errors.Wrap(err1, "cluster CreateTable")
+	}
+
+	// update cache
+	table := c.updateTableCacheLocked(schema, tablePb)
+	return table, nil
+}
+
+func (c *Cluster) loadClusterTopologyLocked(ctx context.Context) error {
+	clusterTopology, err := c.storage.GetClusterTopology(ctx, c.clusterID)
+	if err != nil {
+		return errors.Wrap(err, "cluster loadClusterTopologyLocked")
+	}
+	c.metaData.clusterTopology = clusterTopology
+
+	if c.metaData.clusterTopology == nil {
+		return ErrClusterTopologyNotFound.WithCausef("cluster:%v", c)
+	}
+	shardIDs := make([]uint32, len(c.metaData.clusterTopology.ShardView))
+	shardMap := make(map[uint32][]*clusterpb.Shard)
+
+	for _, shard := range c.metaData.clusterTopology.ShardView {
+		shardMap[shard.Id] = append(shardMap[shard.Id], shard)
+		shardIDs = append(shardIDs, shard.Id)
+	}
+	c.metaData.shardMap = shardMap
+	c.metaData.shardIDs = shardIDs
+	return nil
+}
+
+func (c *Cluster) loadShardTopologyLocked(ctx context.Context) error {
+	topologies, err := c.storage.ListShardTopologies(ctx, c.clusterID, c.metaData.shardIDs)
```

Comment:
Maybe we should move the `c.metaData.shardIDs` as the input param otherwise an inexplicit assumption is introduced here that `loadShardTopologyLocked` should be called after `loadClusterTopologyLocked`.